store/postgres/src/store.rs

```diff
@@ -854,6 +1019,69 @@ impl StoreTrait for Store {
         )
     }
 
+    fn supports_proof_of_indexing<'a>(
+        &'a self,
+        subgraph_id: &'a SubgraphDeploymentId,
+    ) -> DynTryFuture<'a, bool> {
+        self.with_entity_conn(subgraph_id, |conn, cancel| {
+            cancel.check_cancel()?;
+            Ok(conn.supports_proof_of_indexing())
+        })
+        .boxed()
+    }
+
+    fn get_proof_of_indexing<'a>(
+        &'a self,
+        subgraph_id: &'a SubgraphDeploymentId,
+        block_number: u64,
+    ) -> DynTryFuture<'a, Option<ProofOfIndexingDigest>> {
+        let logger = self.logger.cheap_clone();
+
+        self.with_entity_conn(subgraph_id, move |conn, cancel| {
+            cancel.check_cancel()?;
+
+            if !conn.supports_proof_of_indexing() {
+                return Ok(None);
+            }
+
+            let entities = conn
+                .query(
+                    &logger,
+                    EntityCollection::All(vec![POI_OBJECT.to_owned()]),
+                    None,
+                    None,
+                    EntityRange {
+                        first: None,
+                        skip: 0,
+                    },
+                    block_number.try_into().unwrap(),
+                )
+                .map_err(Error::from)?;
+            cancel.check_cancel()?;
+
+            let by_causality_region = entities
+                .into_iter()
+                .map(|e| {
+                    let causality_region = e.id()?;
+                    let digest = match e.get("digest") {
+                        None => Err(format_err!("Entity is missing a digest attribute")),
+                        Some(Value::String(s)) => Ok(s.to_owned()),
+                        _ => Err(format_err!("Entity has non-string digest attribute")),
+                    }?;
+
+                    let poi = ProofOfIndexingDigest(digest);
+
+                    Ok((causality_region, poi))
+                })
+                .collect::<Result<HashMap<_, _>, Error>>()?;
+
+            let hash = stable_hash_with_hasher::<twox_hash::XxHash64, _>(&by_causality_region);
```

Comment:
I checked stable-hash to understand how `HashMap` (by definition unordered) hashes deterministically. I still don't understand it. Could you once more explain how that's guaranteed? I see there are tests for `HashMap` in stable-hash, so this is just for me to understand. 😉

Reply:
The properties we _must_ demonstrate are:

- Determinism: Any value which `PartialEq(a, b) == true` outputs `StableHash(a) == StableHash(b)`.
- Avalanche: Small changes in the input drastically change the output "seemingly randomly"
- Collision resistance: It is not easy to have 2 inputs that hash to the same value.

The algorithm for an unordered collection of *unique* (<- very important to be unique) items is as follows:
```
1: Initialize some binary value (rollup) to all 0 (or some predefined value), and count to 0
2: For each element in the collection:
    a: StableHash it independently of the supplied hash state
    b: XOR the intermediate digest from step a with the rollup
    c: increment count
3: Feed rollup into the StableHash
4: Feed the count into the StableHash
```

### Determinism
We'll take as a given that the StableHash of each element in the sequence has a deterministic stable hash.
Consider the rollup value as an array of bits. Each XOR which operates on the rollup bits operates on each bit independently. Eg: rollup_bits[0] is a function of all of the intermediate_digest_bits[0], and is not a function of any bits[1] or intermediate_digest_bits[1] etc.
Each XOR flips the rollup bit from 0 -> 1 or back if and only if digest_bit is 1. So, the final value of the bit is `count_ones(digest bits) % 2`. The function `count_ones(digest bits)` is independent of the order of `digest bits` and therefore independent of the order of the sequence.
Showing that `rollup` is deterministic was the hard part, it's pretty easy to see that `count` and what we do after is deterministic so that's left as an exercise for the reader.

### Avalanche
This falls out of 2 properties - first that each individual item in the sequence has it's own `StableHash` with the avalanche property, and secondly that XOR acts as a one-time-pad cipher with the probability of each bit being exactly 50%.

### Collision resistance
This falls out of each `StableHash` value being pseudo-unique, there not being duplicate items in the sequence, and including the count.

If there are duplicate items then a hash of `ABA` is the same as the hash of `B` because the XOR cancels.

The count is necessary for any implementation of `StableHasher` which might output `0` for the empty sequence. Such a stable hasher would make it so that `HashSet([])` and `HashSet([0])` collide. I guess this probably isn't necessary in practice.

### Performance

Another property we would _like_ to show is we can get good performance or that the chosen algorithm scales well for performance as compared to other more obvious methods.

The other "obvious" method would have been to sort the sequence and feed the sorted sequence through the hasher.

Sorting takes `O(N*log(N))` instead of `N` but more crucially requires that all the data be loaded and implement `Ord` (ruling out `f64` keys, among others). Although the implementation doesn't make use of this yet, not sorting makes it possible to stream batches of ipfs data sources from the database and update the hash incrementally.

Reply:
While that's a great explanation, I consider using `HashMap` tricky in code where determinism is a requirement. It's easy to iterate over the items without remembering that the order is not deterministic. So we should prefer `BTreeMap` whenever possible, it would be nice to have a `StableHash` implementation for it.

Reply:
@leoyvens It would be nice to have a `StableHash` implementation for `BTreeMap`. I'm not convinced about preferring `BTreeMap` in the general case though. [The docs](https://doc.rust-lang.org/std/collections/index.html#use-a-hashmap-when) have some nice guidelines about when to choose `HashMap` or `BTreeMap` which I find agreeable.

Reply:
Those docs kinda suggest defaulting to `HashMap` if you don't need `BTreeMap`, but I feel the opposite. Not all `HashMap` operations are deterministic. All `BTreeMap` operations are deterministic. In any codebase avoiding accidental non-determinism is good, and in ours I feel `HashMap` is a needless risk.

Reply:
Anyways I'm not saying this PR needs to change, it's just a general note.

Reply:
"any codebase avoiding accidental non-determinism" -> almost all codebases? I mean, it's not like our application is unusual in requiring determinism. So, you're pretty much condemning the use of `HashMap` almost everywhere.

In my experience, this is a bug that bites every developer... _once_. And it's usually a pretty easy fix with consequences not larger than other ordinary bugs. When we consider broadly applying a set of rules as a preventative measure we need to take into account the tradeoffs against the likelihood and severity of the problem manifesting.

It's a pretty subtle issue that I think reasonable people can disagree on.  I'd still choose `HashMap` here again. But, I certainly appreciate the sentiment from which your concern emerges and can respect that you may choose `BTreeMap`.